File: macapp/Sources/PPIDE/Editor/ProjectSearch.swift
import Foundation

/// One search match: a file, a 1-based line number, and the trimmed line text.
struct SearchHit: Identifiable, Hashable {
    let id = UUID()
    let url: URL
    let line: Int
    let preview: String
}

/// Project-wide text search and replace over the files under a root folder.
/// Foundation only; skips hidden files and anything that isn't valid UTF-8 text.
enum ProjectSearch {
    static let maxHits = 500

    static func search(_ query: String, in root: URL) -> [SearchHit] {
        guard !query.isEmpty else { return [] }
        var hits: [SearchHit] = []
        for url in textFiles(under: root) {
            if hits.count >= maxHits { break }
            guard let content = try? String(contentsOf: url, encoding: .utf8) else { continue }
            var lineNo = 0
            content.enumerateLines { line, stop in
                lineNo += 1
                if line.range(of: query, options: .caseInsensitive) != nil {
                    hits.append(SearchHit(url: url, line: lineNo,
                                          preview: line.trimmingCharacters(in: .whitespaces)))
                    if hits.count >= maxHits { stop = true }
                }
            }
        }
        return hits
    }

    /// Replace every (case-sensitive) occurrence of `query` with `replacement` across the
    /// project. Returns the URLs of files that changed.
    @discardableResult
    static func replaceAll(_ query: String, with replacement: String, in root: URL) -> [URL] {
        guard !query.isEmpty else { return [] }
        var changed: [URL] = []
        for url in textFiles(under: root) {
            guard let content = try? String(contentsOf: url, encoding: .utf8),
                  content.contains(query) else { continue }
            let updated = content.replacingOccurrences(of: query, with: replacement)
            if updated != content, (try? updated.write(to: url, atomically: true, encoding: .utf8)) != nil {
                changed.append(url)
            }
        }
        return changed
    }

    private static func textFiles(under root: URL) -> [URL] {
        let fm = FileManager.default
        guard let enumerator = fm.enumerator(
            at: root,
            includingPropertiesForKeys: [.isRegularFileKey],
            options: [.skipsHiddenFiles]
        ) else { return [] }

        var files: [URL] = []
        for case let url as URL in enumerator {
            if (try? url.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile) == true {
                files.append(url)
            }
        }
        return files
    }
}

File: winapp/CodeCrack.App.Core/Search/ProjectSearch.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace CodeCrack.App.Core.Search;

/// One search match: a file path, a 1-based line number, and the trimmed line text.
public sealed record SearchHit(string Path, int Line, string Preview);

/// Project-wide text search and replace under a root folder. Skips hidden files and anything
/// that is not valid UTF-8 text. Search is case-INsensitive; ReplaceAll is case-SENSITIVE —
/// the asymmetry is deliberate and mirrors the mac app.
public static class ProjectSearch
{
    public const int MaxHits = 500;

    // Strict decoder: throws on invalid UTF-8 so we can skip binary files.
    private static readonly Encoding StrictUtf8 =
        new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);

    public static IReadOnlyList<SearchHit> Search(string query, string root)
    {
        var hits = new List<SearchHit>();
        if (string.IsNullOrEmpty(query)) return hits;

        foreach (var path in TextFiles(root))
        {
            if (hits.Count >= MaxHits) break;
            if (!TryReadUtf8(path, out var content)) continue;

            var line = 0;
            foreach (var raw in content.Split('\n'))
            {
                line++;
                var text = raw.TrimEnd('\r');
                if (text.IndexOf(query, StringComparison.OrdinalIgnoreCase) >= 0)
                {
                    hits.Add(new SearchHit(path, line, text.Trim()));
                    if (hits.Count >= MaxHits) break;
                }
            }
        }
        return hits;
    }

    /// Replace every case-sensitive occurrence of query with replacement across the project.
    /// Returns the paths of files that actually changed.
    public static IReadOnlyList<string> ReplaceAll(string query, string replacement, string root)
    {
        var changed = new List<string>();
        if (string.IsNullOrEmpty(query)) return changed;

        foreach (var path in TextFiles(root))
        {
            if (!TryReadUtf8(path, out var content)) continue;
            if (!content.Contains(query, StringComparison.Ordinal)) continue;

            var updated = content.Replace(query, replacement, StringComparison.Ordinal);
            if (updated == content) continue;
            File.WriteAllText(path, updated, StrictUtf8); // UTF-8, no BOM
            changed.Add(path);
        }
        return changed;
    }

    private static IEnumerable<string> TextFiles(string root)
    {
        if (!Directory.Exists(root)) yield break;
        var opts = new EnumerationOptions
        {
            RecurseSubdirectories = true,
            AttributesToSkip = FileAttributes.Hidden | FileAttributes.System,
            IgnoreInaccessible = true
        };
        foreach (var path in Directory.EnumerateFiles(root, "*", opts))
        {
            if (Path.GetFileName(path).StartsWith('.')) continue; // dotfiles
            yield return path;
        }
    }

    private static bool TryReadUtf8(string path, out string content)
    {
        try { content = File.ReadAllText(path, StrictUtf8); return true; }
        catch { content = ""; return false; } // DecoderFallbackException on binary, or IO error
    }
}
